### DISCUSSION TITLE
Requirements and the nature of Power-Fx?

### DISCUSSION THREAD

**Discussion by mikeschinkel** [microsoft/Power-Fx#232] [Q&A] [answered]
What are the requirements for using/working on PowerFx?   Windows, or does it work on macOS?

Also, is this being continued, or was it just a PoC and paid, hosted PowerApps is taking over as how PowerFx will move forward?

**Comment by MikeStall** [comment]
Power Fx is very much an ongoing active project!  You can see the frequency of updates in the commit history (https://github.com/microsoft/Power-Fx/commits/main) . 

Power Fx is built with .net core and should run cross platform. Although the majority of developers do use Windows. 

We do welcome contributions - first file an issue, get consensus, and then submit a PR.

**Reply by mikeschinkel** [reply]
Hi @MikeStall — Thanks for such a fast response.

However, I think your answer just expanded the number of questions I have. Let me try to elaborate as I am current very confused by it all. And sorry in advance if some of these questions overlap as it's hard to get my thoughts clear about what to ask.

_(BTW, I am a professional software developer, currently specializing in Go but with experience in PHP/MySQL and also ASP/IIS/SqlServer long ago.)_

1. What exactly _is_ Power-Fx?  Is PowerFX a programming language like C# is a programming language, or it is a library for use in C# that includes an API that interprets Excel-like formulas?

1. Is there a command-line compiler for Power-Fx specific YAML files?  If yes, I can't seem to find it. If no, how does one process YAML files, and what is their standard extension, e.g. `.pfx` or something else?

1. How would one create the simplest possible "hello world" app?

1. Can Power-Fx generate standalone executables? If yes, can it generate ones for cross-platform? If no, does it require a runtime for local execution and/or is it limited to running via a web browser/web server combination?

1. When I look at the "language" documentation it appears as it it is more of a documentation about a programmable feature of the online Power Apps, and not a programming language per se.  Is that analysis wrong?

1. Does Power-Fx require Visual Studio 2019? I can't seem to find any way to just load it in a text editor and work with it. 

1. I use the JetBrains IDEs primarily and I opened Power-Fx in [Rider ](https://www.jetbrains.com/rider/) but was at a complete loss for how to compile it or turn it into something workable. So I installed Visual Studio for Mac and clicked the build button and as far as I can tell it generated a `Microsoft.PowerFx.Core.0.3.0-local.nupkg` in the `outputpackages` directory. From googling that is a package.  How would I use that package?

Sorry if I am asking so many questions. I am interested in Power-Fx because of a family member who uses Excel extensively recently lamented to me that he wished he had _"learned how to program"_ when he was younger so he could package up programs like a developer who uses a compiler to create applications. 

So my motivation is to maybe learn Power-Fx it so I can teach it to him so he can realize that he actually still can learn how to program.

Thanks in advance for the help.

**Comment by MikeStall** [comment, marked as answer]
See here for more on Power Fx: https://docs.microsoft.com/en-us/power-platform/power-fx/overview
Power Fx is a low-code customization language - that means it's hosted inside other products (like PowerApps) and not a standalone language.  So to "learn power fx", just learn Power Apps:  https://powerapps.microsoft.com/en-us/landing/developer-plan

To answer your questions:
1. We are sharing power fx as a standalone project on github so that other products and integrate it in the same way that powerapps does. 
2. No command line.  
3. Via PowerApps. 
4. PowerFx isn't a standalone language - but we do have a sample repl , https://github.com/microsoft/power-fx-host-samples/tree/main/Samples/ConsoleREPL
5. Correct. PowerApps is the biggest consumer of Power Fx, but other products can integrate it too. 
6. To build the power fx compiler itself, use Visual Studio. To consume power fx as an end-user, use a product like PowerApps.
7. See the samples (from 4)

**Reply by mikeschinkel** [reply]
Hey @MikeStall,

Thank you again for answering my questions.

Assuming you are not interested in feedback from prospective users, feel free to ignore the rest.

But if you are open to feedback it is disheartening for me to learn that Power-Fx does not appear to be a standalone language. When I first learned of Power-Fx I assumed it could be _"the modern equivalent of pre-.NET Visual Basic that would empower legions of [occupational programmers](https://mikeschinkel.com/2007/will-microsoft-meet-occupational-programmers-needs/) that  I have so long felt Microsoft and everyone else have been ignoring."_ 

Unless I am wrong, Power-Fix seems not so much a language but a feature of (an)other tool(s). Which is definitely not a modern pre-.NET Visual Basic, sadly, for me at least. 

If you don't mind, can you answer a few more questions?

1. So the open-source Power-Fx is really just the language parser and evaluator, and does not provide any of the UI nor the framework that enables the event-driven recalculation. Is that correct?

2. So there is not a vision for YAML scripts to be files on disk that are compiled; YAML is merely something that someone would pasted into a text box in some app that hosts Power-Fx where a YAML parser would feed each of the formulas into Power-Fx. Is that correct?

3. Currently Power Apps hosts Power-Fx. Are there any other products from other companies currently available that integrate Power-Fx? If yes, can you enumerate them?

4. Are there any other companies you are aware of working on products not yet available that will integrate Power-Fx? If yes, is there anything you can say about them?

In closing, based on the explosive growth in Visual Basic that fueled my company-at-the-time VBxtras I could envision massive growth for Power-Fx if is enabled occupational programmers in a similar way, but not if just a library for use in an app that needs to implement all the event-driven UI. So I am really interested to learn if it could be that.

**Comment by MikeStall** [comment]
What you're looking for is **Power Apps** - that includes the ability to create visual apps over data, similar to what VB/VBA did. 

Power-fx is "just" the language inside of Power Apps, and rehosted so that other things can use it too, and Power Apps is having explosive growth. For example, Dataverse allows defining columns via power fx expressions (https://powerapps.microsoft.com/en-us/blog/power-fx-formula-columns-in-dataverse/),  and there's also a commanding integration (https://docs.microsoft.com/en-us/powerapps/maker/model-driven-apps/commanding-use-powerfx).  3rd parties are also integrating power fx into their products (see other examples in this discussion group). 
This is a common pattern for languages. C# is "just" a language, but needs to be used with other libraries and frameworks (such as ASP.Net or WPF) to do something useful. 

1. Yes, although there are other pieces (like PowerApps, Dataverse, etc) that provide those other pieces. 
2. not quite.  There is a yaml representation for power apps : https://powerapps.microsoft.com/en-us/blog/source-code-files-for-canvas-apps/ 
3. Yes, see above. 
4. Yes and no.  ;)

**Reply by mikeschinkel** [reply]
Hey @MikeStall — 

Also, I submitted a few PRs to add your video and article links to the Github repo so that for other people in the future with similar questions to mine will discover there answers quicker and easier than having to ask a bunch of questions like I did! 😁

- https://github.com/microsoft/Power-Fx/pull/234
- https://github.com/microsoft/Power-Fx/pull/235

**Comment by mikeschinkel** [comment]
Hi @MikeStall —

Thank you again for the detailed reply

> _"What you're looking for is **Power Apps** - that includes the ability to create visual apps over data, similar to what VB/VBA did."_

Actually, not really. Power Apps — unless I misunderstood — cannot build standalone executables nor is it open-source. 

But I do get your point about it having the other functionality I was expecting to find here.

> https://powerapps.microsoft.com/en-us/blog/source-code-files-for-canvas-apps/

This was helpful, thanks.

Further, after looking through some old discussions I came across [**this video**](https://youtu.be/ik6k89WNjuk) with you and Greg which was super helpful. You really should **showcase this video on the homepage** of this Github repo. Had I found that first I would have had many fewer questions!  

### Summary for posterity
So my current takeaway just for posterity is that Power-Fx isn't a _"programming language"_ in the way most programmers think about programming languages — C, C#, Python, Ruby, PHP, Go, Rust, PowerShell, Bash, AWK, etc. — but instead a **formula-based calculation engine** that has its own **mini-language** _(interestingly, (ironically?) when I googled that term I found [this](https://handsontable.github.io/hyperformula/#what-is-hyperformula).)_ 

Further, I do now see where Power-Fx can really shine when embedded in other end-user applications like with [Acumatica](https://www.youtube.com/watch?v=ik6k89WNjuk&t=1025s) where it empowers end-users to configure an app using complex logic in a manner that would usually require custom programming. _(That video of Acumatica reminded me an awful lot of configuring Salesforces, btw.)_ But on its own, it's not sufficient to create an end-user application that someone might have started creating in Excel.

And of course for those citizen developers — what I called _"[_occupational programmers_](https://mikeschinkel.com/2007/will-microsoft-meet-occupational-programmers-needs/)"_ 15 years ago — who can maintain a Power Apps account funded by their business needs then Power-Fx can provide powerful _"low-code"_ app-creation capability. But that app-creation capability is not what has been open-sourced, only the formula-based calculation engine has.

So, is that a fair summary?
